distributed/core.py

```diff
@@ -949,18 +949,15 @@ def remove(self, addr):
                 IOLoop.current().add_callback(comm.close)
                 self.semaphore.release()
 
-    def close(self):
+    async def close(self):
         """
         Close all communications abruptly.
         """
-        for comms in self.available.values():
-            for comm in comms:
-                comm.abort()
+        for d in [self.available, self.occupied]:
+            comms = [comm for comms in d.values() for comm in comms]
+            await asyncio.gather(*[comm.close() for comm in comms])
```

Comment:
Currently, this does not seem to release the semaphore in case `comm.close` raises. I think one can make i safer by using `return_exceptions=True` in `gather`.
(OTOH, if `close` really means this instance should never be used again, it does not matter, of course. However, in that case, it would be nice to be more explicit about that and fail in case one tries to use a "closed" pool).

Reply:
Thanks @byjott !  I've added the `return_exceptions=True` keyword, which I agree is likely a good idea.  In practice you're correct in guessing that this is only used at process shutdown, but it doesn't hurt.

I haven't yet added any sort of check on reuse of a closed connection pool.  If it's ok I'd like to pass on this until it comes up (which I suspect it won't)